### PR TITLE
[WIP] Add support for Int64/Int32 -> Int8 coersions

### LINKED ISSUE #47: Fix Cmm narrowing operators for `Int8#`/`Word8#`
Note: `Int8#`/`Word8#` isn't widely used in the boot libs yet, so this is a medium-priority issue and will be checked only after #44 

Cmm has narrowing operators `MO_SS_Conv`/`MO_UU_Conv` which narrows a 64-bit `Int#`/`Word#` to 32-bit or lower, while still using a 64-bit local/global register for storage. The codegen already handles 64-to-32 narrowing, but the newly added `Int8#`/`Word8#` requires more aggressive 64-to-8 narrowing. Note that wasm doesn't have native opcodes for this purpose, so we need a pair of store/load opcodes here. When we later `Worker`ify the runtime, we also need extra care not to introduce a race condition since it involves a pinned temporary memory region.

### PR DESCRIPTION
**Pull request by bollu**
Fixes #47

### CODE REVIEW

**Comment by bollu** [comment]
I get the error:
```
[wasm-validator error in function $wrapI64ToI8] unexpected true: if block is not returning a value, final element should not flow out a value, on 
[none] [none] (block
 [none] (i64.store align=1
  [i32] (i32.wrap_i64
   [i64] (i64.const 9007160600037488)
  )
  [i64] (local.get $0)
 )
 [none] (local.set $3
  [i64] (i64.load align=1
   [i32] (i32.wrap_i64
    [i64] (i64.const 9007160600037488)
   )
  )
 )
 [i64] (local.get $3)
) ;; end block
```

which I don't understand (so I'm reading the `binaryen` docs on some explanation for their IR).
why does it think that the block is not returning a value? there is a `[i64] (local.get $0)`

File: asterius/src/Asterius/Builtins.hs [hunk 1]
```diff
@@ -1538,3 +1552,22 @@ trapStoreF64Function _ =
 
 offset_StgTSO_StgStack :: Int
 offset_StgTSO_StgStack = 8 * roundup_bytes_to_words sizeof_StgTSO
+
+
+wrapI64ToI8 _ =
+  let v = runEDSL [] $ do
```
Comment (TerrorJack) [inline asterius/src/Asterius/Builtins.hs]: This looks fishy. `runEDSL [I32]` maybe?

File: asterius/src/Asterius/Builtins.hs [hunk 2]
```diff
@@ -1538,3 +1552,22 @@ trapStoreF64Function _ =
...
+        setReturnTypes [I32]
+        x <- param I64
+        -- TODO: add __i64_slot into EDSL.hs
+        storeI64 (symbol "__asterius_i64_slot") 0 x
+        v <- i64Local $ loadI64 (symbol "__asterius_i64_slot") 0
```
Comment (TerrorJack) [inline asterius/src/Asterius/Builtins.hs]: `loadI8` maybe?

Comment (bollu) [reply asterius/src/Asterius/Builtins.hs]: Oh god, I'm stupid. Yes, this should be the fix.

File: asterius/src/Asterius/CodeGen.hs [hunk 3]
```diff
@@ -311,10 +312,29 @@ marshalCmmHomoConvMachOp ::
...
+    traceM $ "in marshal: " <> show w0 <> " -> " <> show w1
+    -- TODO: we are assuming that always wrap to I8, it could be I16 as well
+    (xe, _) <- marshalCmmExpr x
+    let c = Call
+              { target = if w0 == GHC.W32 then "wrapI32ToI8" else "wrapI64ToI8"
+              , operands = [xe]
+              , callReturnTypes = [I32]
+              }
+
+    let c' = CallImport
```
Comment (TerrorJack) [inline asterius/src/Asterius/CodeGen.hs]: What's `c'` used for?

Comment (bollu) [reply asterius/src/Asterius/CodeGen.hs]: Dead code.

**Comment by bollu** [comment]
Mh, nope, the changes don't help. Still get the output:

```
(1,63)
(1,255)
(1,255)
(1,255)
199990000
```

**Comment by bollu** [comment]
```
╭─bollu@strangeattractor ~/work/asterius ‹may-10-add-support-for-int8› 
╰─$ git rev-parse HEAD

1a18f00c41977b26e28b5601cbc320852360e24e
╭─bollu@strangeattractor ~/work/asterius ‹may-10-add-support-for-int8› 
╰─$ stack test asterius:bytearraymini  --test-arguments="--binaryen --output-ir"
asterius-0.0.1: test (suite: bytearraymini, args: --binaryen --output-ir)

[INFO] Compiling test/bytearraymini/bytearraymini.hs to WebAssembly
Function {functionType = FunctionType {paramTypes = [I64], returnTypes = [I32]}, varTypes = [I32,I32], body = Block {name = "", bodys = [Store {bytes = 8, offset = 0, ptr = Unary {unaryOp = WrapInt64, operand0 = Symbol {unresolvedSymbol = "__asterius_i64_slot", symbolOffset = 0}}, value = GetLocal {index = 0, valueType = I64}, valueType = I64},Load {signed = False, bytes = 1, offset = 0, valueType = I32, ptr = Unary {unaryOp = WrapInt64, operand0 = Symbol {unresolvedSymbol = "__asterius_i64_slot", symbolOffset = 0}}}], blockReturnTypes = [I32]}}
Function {functionType = FunctionType {paramTypes = [I64], returnTypes = [I32]}, varTypes = [I32,I32], body = Block {name = "", bodys = [Store {bytes = 8, offset = 0, ptr = Unary {unaryOp = WrapInt64, operand0 = Symbol {unresolvedSymbol = "__asterius_i64_slot", symbolOffset = 0}}, value = GetLocal {index = 0, valueType = I64}, valueType = I64},Load {signed = False, bytes = 2, offset = 0, valueType = I32, ptr = Unary {unaryOp = WrapInt64, operand0 = Symbol {unresolvedSymbol = "__asterius_i64_slot", symbolOffset = 0}}}], blockReturnTypes = [I32]}}
[INFO] Serializing linked IR to "test/bytearraymini/bytearraymini.bin"
[INFO] Converting linked IR to binaryen IR
[INFO] Validating binaryen IR
[INFO] Writing WebAssembly binary to "test/bytearraymini/bytearraymini.wasm"
[INFO] Writing re-parsed wasm-toolkit IR to "test/bytearraymini/bytearraymini.binaryen.txt"
[INFO] Writing JavaScript runtime settings to "test/bytearraymini/rts.settings.mjs"
[INFO] Writing JavaScript runtime modules to "test/bytearraymini"
[INFO] Writing JavaScript loader module to "test/bytearraymini/bytearraymini.wasm.mjs"
[INFO] Writing JavaScript lib module to "test/bytearraymini/bytearraymini.lib.mjs"
[INFO] Writing JavaScript entry module to "test/bytearraymini/bytearraymini.mjs"
[INFO] Running test/bytearraymini/bytearraymini.mjs
(node:31779) ExperimentalWarning: The ESM module loader is experimental.
255


asterius-0.0.1: Test suite bytearraymini passed
╭─bollu@strangeattractor ~/work/asterius ‹may-10-add-support-for-int8› 
╰─$ runghc asterius/test/bytearraymini/bytearraymini.hs 
-1
```

**Comment by bollu** [comment]
Not sure how useful the module dump is, it's _huge_ (like, `3.4MB`), but [here's the gist](https://gist.github.com/bollu/271441416f178010ab56221ba46fdd80)

**Comment by bollu** [comment]
latest gist: https://gist.github.com/bollu/a4ebfc0eab259de1ca00b70edd247f47

**Comment by bollu** [comment]
This currently dies at:

**Comment by bollu** [comment]
Oh hm, even the `fib` testcase breaks:
```
╭─bollu@strangeattractor ~/work/asterius ‹may-10-add-support-for-int8› 
╰─$ ASTERIUS_DEBUG=1 stack build  && rm -rf  asterius/test/bytearraymini && git checkout -- asterius/test/bytearraymini &&  stack test asterius:fib --test-arguments="--binaryen --output-ir"          
asterius-0.0.1: test (suite: fib, args: --binaryen --output-ir)

[INFO] Compiling test/fib/fib.hs to WebAssembly
Function {functionType = FunctionType {paramTypes = [I64], returnTypes = [I32]}, varTypes = [I32,I32], body = Block {name = "", bodys = [Store {bytes = 8, offset = 0, ptr = Unary {unaryOp = WrapInt64, operand0 = Symbol {unresolvedSymbol = "__asterius_i64_slot", symbolOffset = 0}}, value = GetLocal {index = 0, valueType = I64}, valueType = I64},Load {signed = False, bytes = 1, offset = 0, valueType = I32, ptr = Unary {unaryOp = WrapInt64, operand0 = Symbol {unresolvedSymbol = "__asterius_i64_slot", symbolOffset = 0}}}], blockReturnTypes = [I32]}}
Function {functionType = FunctionType {paramTypes = [I64], returnTypes = [I32]}, varTypes = [I32,I32], body = Block {name = "", bodys = [Store {bytes = 8, offset = 0, ptr = Unary {unaryOp = WrapInt64, operand0 = Symbol {unresolvedSymbol = "__asterius_i64_slot", symbolOffset = 0}}, value = GetLocal {index = 0, valueType = I64}, valueType = I64},Load {signed = False, bytes = 2, offset = 0, valueType = I32, ptr = Unary {unaryOp = WrapInt64, operand0 = Symbol {unresolvedSymbol = "__asterius_i64_slot", symbolOffset = 0}}}], blockReturnTypes = [I32]}}
[INFO] Serializing linked IR to "test/fib/fib.bin"
[INFO] Converting linked IR to binaryen IR
[INFO] Validating binaryen IR
[wasm-validator error in function $base_GHCziForeign_.Lc8JLq] unexpected false: expected i32 operation to touch 1, 2, or 4 bytes, on 
[unreachable] (i32.store align=1
 [i32] (i32.wrap_i64
  [i64] (i64.load offset=872 align=1
   [i32] (i32.wrap_i64
    [i64] (i64.const 9007160600035328)
   )
  )
 )
 [unreachable] (unreachable)
)
[wasm-validator error in function $base_GHCziIOziEncodingziUTF8_.Lc8AvQ] unexpected false: expected i32 operation to touch 1, 2, or 4 bytes, on 
[unreachable] (i32.store align=1
 [i32] (i32.wrap_i64
  [i64] (i64.load offset=872 align=1
   [i32] (i32.wrap_i64
    [i64] (i64.const 9007160600035328)
   )
  )
 )
 [unreachable] (unreachable)
)
ahc-link: user error ([ERROR] binaryen validation failed)
fib: callProcess: ahc-link "--input-hs" "test/fib/fib.hs" "--run" "--binaryen" "--output-ir" (exit 1): failed

asterius-0.0.1: Test suite fib failed
Test suite failure for package asterius-0.0.1
    fib:  exited with: ExitFailure 1
Logs printed to console
```

Time to see what I changed so drastically :)

**Comment by bollu** [comment]
Closed in favor of #114

**Pull request closed without merging**